fingerprint/infra/scanner/src/main/java/com/simprints/fingerprint/infra/scanner/v2/channel/MessageChannel.kt

```diff
@@ -2,19 +2,35 @@ package com.simprints.fingerprint.infra.scanner.v2.channel
 
 import com.simprints.fingerprint.infra.scanner.v2.incoming.common.MessageInputStream
 import com.simprints.fingerprint.infra.scanner.v2.outgoing.common.MessageOutputStream
+import com.simprints.fingerprint.infra.scanner.v2.scanner.errorhandler.ResponseErrorHandler
+import com.simprints.fingerprint.infra.scanner.v2.scanner.errorhandler.ResponseErrorHandlingStrategy
 import io.reactivex.Flowable
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
 import java.io.OutputStream
 
 /**
- * Base class to hold onto pairs of corresponding [MessageInputStream]s and [MessageOutputStream]s
- * for easier sending and receiving.
+ * Represents a message channel with incoming and outgoing streams,
+ * providing thread-safe and coroutine-friendly operations.
  *
- * It was found to be better to start observing the response at the same time/slightly before
- * sending the command as sometimes the Vero would respond too quickly.
+ * @param I The type of the input stream, must extend [MessageInputStream].
+ * @param O The type of the output stream, must extend [MessageOutputStream].
+ * @property incoming The incoming stream for reading data.
+ * @property outgoing The outgoing stream for writing data.
+ * @property readWriteLock A mutex for synchronizing read/write access.
+ * @property responseErrorHandler Handles errors during execution.
+ * @property dispatcher The coroutine dispatcher for executing tasks.
  */
 abstract class MessageChannel<I : MessageInputStream, O : MessageOutputStream<*>>(
     val incoming: I,
-    val outgoing: O
+    val outgoing: O,
+    val dispatcher: CoroutineDispatcher,
+    val readWriteLock: Mutex = Mutex(),
```

Comment:
I just want to clarify - is there one lock per channel? Is there a risk of interference between channels (assuming they use the same BT connection on the low level)

Reply:
Each channel has its own mutex. In the app, there is always only one active channel at a time. For instance, during the normal capture flow, we transition from the root (setting up the connection and turning un20 on) to the main mode (capturing the fingerprint).